File: problem2.cpp
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
public:
    string decodeString(string s) {

        string ans="";
        int num=0;
        stack<pair<string,int>>stk;
        for(int i=0;i<s.length();i++)
        {
            if(isdigit(s[i]))
                num=(num*10)+(s[i]-48);
           else if(isalpha(s[i]))
                ans+=s[i];
           else if(s[i]=='[')
            {
                stk.push({ans,num});
                ans="";
                num=0;
            }
            else if(s[i]==']')
            {
                string tmp=ans;
                for(int i=1;i<stk.top().second;i++)                
                   tmp+=ans;
                ans=stk.top().first+tmp;
                stk.pop();
                
            }
                
        }
     
        return ans;
    }
};

File: PRoblem2.py
# Time Complexity : O(maxK* countK)
# Space Complexity : O(m+n)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

class Solution(object):
    def decodeString(self, s):
        """
        :type s: str
        :rtype: str
        """
        stack = []; curNum = 0; curString = ''
        for c in s:
            if c == '[':
                stack.append(curString)
                stack.append(curNum)
                curString = ''
                curNum = 0
            elif c == ']':
                num = stack.pop()
                prevString = stack.pop()
                curString = prevString + num*curString
            elif c.isdigit():
                curNum = curNum*10 + int(c)
            else:
                curString += c
        return curString